aptos-move/framework/aptos-framework/sources/nonce_validation.move

```diff
@@ -0,0 +1,236 @@
+        txn_expiration_time: u64,
+        sender_address: address,
+        nonce: u64,
+    }
+
+    struct NonceKey has copy, drop, store {
+        sender_address: address,
+        nonce: u64,
+    }
+
+    public(friend) fun initialize(aptos_framework: &signer) {
+        initialize_nonce_table(aptos_framework);
+    }
+
+    public entry fun initialize_nonce_table(aptos_framework: &signer) {
+        if (!exists<NonceHistory>(@aptos_framework)) {
+            let table = table::new();
+            let nonce_history = NonceHistory {
+                nonce_table: table,
+                next_key: 0,
+            };
+            move_to<NonceHistory>(aptos_framework, nonce_history);
+        };
+    }
+
+    fun empty_bucket(): Bucket {
+        let bucket = Bucket {
+            nonces_ordered_by_exp_time: big_ordered_map::new_with_config(0, 0, true),
+            nonce_to_exp_time_map: big_ordered_map::new_with_config(0, 0, true),
+        };
+
+        // Initiating big ordered maps with 5 pre-allocated storage slots.
+        // (expiration time, address, nonce) is together 48 bytes.
+        // A 4 KB storage slot can store 80+ such tuples.
+        // The 5 slots should be more than enough for the current use case.
+        bucket.nonces_ordered_by_exp_time.allocate_spare_slots(5);
+        bucket.nonce_to_exp_time_map.allocate_spare_slots(5);
+        bucket
+    }
+
+    // This method is used to prefill the nonce_table with empty buckets one by one.
+    public entry fun add_nonce_buckets(count: u64) acquires NonceHistory {
+        assert!(exists<NonceHistory>(@aptos_framework), error::invalid_state(E_NONCE_HISTORY_DOES_NOT_EXIST));
+        let nonce_history = &mut NonceHistory[@aptos_framework];
+        for (i in 0..count) {
+            if (nonce_history.next_key <= NUM_BUCKETS) {
+                if (!nonce_history.nonce_table.contains(nonce_history.next_key)) {
+                    nonce_history.nonce_table.add(
+                        nonce_history.next_key,
+                        empty_bucket()
+                    );
+                };
+                nonce_history.next_key = nonce_history.next_key + 1;
+            }
+        }
+    }
+
+    // Returns true if the input (address, nonce) pair doesn't exist in the nonce history, and inserted into nonce history successfully.
+    // Returns false if the input (address, nonce) pair already exists in the nonce history.
+    public(friend) fun check_and_insert_nonce(
+        sender_address: address,
+        nonce: u64,
+        txn_expiration_time: u64,
+    ): bool acquires NonceHistory {
+        assert!(exists<NonceHistory>(@aptos_framework), error::invalid_state(E_NONCE_HISTORY_DOES_NOT_EXIST));
+        let nonce_history = &mut NonceHistory[@aptos_framework];
+        let nonce_key = NonceKey {
+            sender_address,
+            nonce,
+        };
+        let nonce_key_with_exp_time = NonceKeyWithExpTime {
+            txn_expiration_time,
```

Comment:
given the check of `current_time - MAX_EXPIRATION_TIME < txn_expiration_time < current_time`, maybe it will be more consistent that this is insertion time (i.e. current_time) instead ?

that way rule is that nonce cannot repeat in 65s window exactly

Reply:
Are we sure we want to override the user specified transaction expiration time with `insertion time + MAX_EXPIRATION_TIME`?

Reply:
And if we override the expiration time with `insertion time + MAX_EXPIRATION_TIME`, all the time the transaction spent in previous stages (mempool, qs, consensus, block preparation, execution pool) doesn't count towards the transaction expiration time. The timer starts when the transaction is processed by prologue. Is that fine with us?